hngx-org/authentication_service
# Auth Controller Optimization and email notification

🚀 Refactoring for AuthController Optimization

### Review Summary:
I've conducted a comprehensive review of the AuthController codebase. The focus was on optimizing database interactions by minimizing queries, utilizing indexing, incrementally completing pending critical features and calling the **email service API** to send email notification

[Linear issue](https://linear.app/zuri-project-backend/issue/AUT-26/za-024)
### Controllers

1. Reset password
2. Verify email
3. Forgot Password

**Optimization Strategies:**
Database Efficiency: I've worked on minimizing database queries and optimizing interactions using appropriate indexing strategies. This will significantly enhance the application's performance and responsiveness.
Pending Feature Completion: I've prioritized completing critical pending features incrementally. This ensures a comprehensive and robust solution for our users.

### Goals Achieved:
**Code Quality:** The refactoring efforts have significantly enhanced the overall code quality, making it more maintainable and readable.
**Performance Boost:** Database optimizations and feature completions will lead to a noticeable improvement in the application's performance and user experience.

### How to test
To test the forgotPassword function, you'll need to simulate an HTTP request, typically through a testing framework like Mocha, Jest, or using a tool like Postman for manual testing. Below are the steps to test this function manually using Postman:

Prerequisites:

Ensure the server is running and accessible.
Set up environment variables like process.env.JWT_SECRET, process.env.FRONT_END_URL, etc., in your development environment or Postman environment variables.
Testing with Postman:

1. POST Request:
### **Forgot Password**

In Postman, create a new request and set the HTTP method to POST.
Set the request URL to the endpoint where this function is hosted (e.g., http://localhost:3000/api/forgot-password).
Set the request headers if required (Content-Type, Authorization, etc.).
In the request body, provide valid JSON data according to the forgotPasswordSchema.
 For example:
`{
  "email": "user@example.com"
}
`

2. Send Request:

Click the "Send" button in Postman to send the POST request to the server.

3. Expected Responses:
If the request is successful and the email exists in the database and is verified, you should receive a response with a status of 200 and a JSON object like:

`
{
  "success": true,
  "message": "Password reset link sent successfully."
}
`
### **Reset Password**

**Testing Instructions:**

To test the password reset functionality with newPassword and confirmPassword fields, follow these steps:

Setup:
**Url:** https://auth.akuya.tech/
**Method**: Patch
Ensure the server is running and accessible.
Set up environment variables like process.env.JWT_SECRET and other required variables.

`{
  "newPassword": "newPassword123",
  "confirmPassword": "newPassword123"
}
`
### Next Steps:
I will be providing regular progress updates and maintaining comprehensive documentation throughout the optimization process. Please feel free to reach out if you have any questions or specific areas you'd like me to focus on.

Thank you for your attention to this optimization effort! Looking forward to your feedback.
![Uploading Reset-Password-gaiyaobed94-gmail-com-Gmail.png…]()

### Some screenshot
![Screenshot (3)](https://github.com/hngx-org/authentication_service/assets/95748043/04c2dd43-54ee-4b8f-b1e1-256a5a5b19d0)
![Uploading Screenshot (6).png…]()
![Reset-Password-gaiyaobed94-gmail-com-Gmail](https://github.com/hngx-org/authentication_service/assets/95748043/f2a2ff09-9f30-4aad-802f-e4c6ee8ece13)
